wagtail/permission_policies/base.py

```diff
@@ -128,6 +128,37 @@ def users_with_any_permission_for_instance(self, actions, instance):
     def users_with_permission_for_instance(self, action, instance):
         return self.users_with_any_permission_for_instance([action], instance)
 
+    def _update_instance_annotated_permissions(self, instance, update_dict):
+        """
+        Update the instance.annotated_permissions dictionary with the given update_dict.
+        """
+        annotation = getattr(instance, "annotated_permissions", {})
+        annotation.update(update_dict)
+        instance.annotated_permissions = annotation
+
+    def annotate_with_permissions(self, queryset, user, actions):
+        """
+        Annotate each instance in the given queryset with a annotated_permissions
+        dictionary that maps each action in the given list of actions to a
+        boolean value indicating whether the given user has permission to
+        perform that action.
+
+        Subclasses may override this method to provide a more efficient
+        implementation, e.g. by performing a single database query to
+        determine the permissions for all instances in the queryset.
+        """
+        for instance in queryset:
+            self._update_instance_annotated_permissions(
+                instance,
+                {
+                    action: self.user_has_permission_for_instance(
+                        user, action, instance
+                    )
+                    for action in actions
+                },
+            )
```

Comment:
Just for context as I forgot to leave a comment in my original PR:

This code is used as-is for `ModelPermissionPolicy` (which is used in snippets' `IndexView`). At first glance, it might look like this could trigger a database query for each instance. However, `ModelPermissionPolicy.user_has_permission_for_instance()` makes use of `user.has_perm()` which in turn will use Django's permission cache for the model, so this doesn't use any queries.

If we were to query the `Permission` model ourselves, it would need an extra query compared to what we're doing here.